src/Main.cpp

```diff
@@ -125,188 +75,681 @@ struct ClickControlSystem
+		F32 fnx = -fsignf(movement.velocity.x)*u*n; /* Friction force. */
+		F32 fny = -fsignf(movement.velocity.y)*u*n;
+		F32 fx  = fnx+movement.force.x;             /* Total force. */
+		F32 fy  = fny+movement.force.y;
+
+		for (auto&& [entity2, movement2, pos2] : access_storage().each())
+		{
+			F32 dx, dy;
+			F32 collision_width = movement.collision_box_width + movement2.collision_box_width;
+			F32 collision_height = movement.collision_box_height + movement2.collision_box_height;
 
-	void on_start() override
-	{
-		TextRender::init();
-		level_manager = LevelManager();
-		level_manager.load_level("Levels/level1.txt");
+			if (entity <= entity2)
+			{
+				continue;
+			}
+
+			dx = pos2.xy.x - pos.xy.x;
+			dy = pos2.xy.y - pos.xy.y;
+
+			if (fabsf(dx) < collision_width / 2.0f && fabsf(dy) < collision_height / 2.0f)
+			{
+				if (fabsf(dx) > fabsf(dy))
+				{
+					if (movement.velocity_max.x > 0 || movement.velocity_max.y > 0)
+					{
+						if (dx > 0.0f)
+						{
+							pos.xy.x -= collision_width / 2.0f - dx;
+							/*movement2.velocity.x -= (fx/m)*Time::delta_time();*/
+						}
+						else
+						{
+							pos.xy.x += collision_width / 2.0f + dx;
+							/*movement2.velocity.x += (fx/m)*Time::delta_time();*/
+						}
+					}
+					else
+					{
+						if (dx > 0.0f)
+						{
+							pos2.xy.x += collision_width / 2.0f - dx;
+							/*movement.velocity.x -= (fx/m)*Time::delta_time();*/
+						}
+						else
+						{
+							pos2.xy.x -= collision_width / 2.0f + dx;
+							/*movement.velocity.x -= (fx/m)*Time::delta_time();*/
+						}
+					}
 
-		for(U32 i = 0; i < TILE_ROWS; i++)
+				}
+				else
+				{
+					if (movement.velocity_max.x > 0 || movement.velocity_max.y > 0)
+					{
+						if (dy > 0.0f)
+						{
+							pos.xy.y -= collision_height / 2.0f - dy;
+							/*movement2.velocity.x -= (fx/m)*Time::delta_time();*/
+						}
+						else
+						{
+							pos.xy.y += collision_height / 2.0f + dy;
+							/*movement2.velocity.x += (fx/m)*Time::delta_time();*/
+						}
+					}
+					else
+					{
+						if (dy > 0.0f)
+						{
+							pos2.xy.y += collision_height / 2.0f - dy;
+							movement.velocity.x += (fx/m)*Time::delta_time();
+						}
+						else
+						{
+							pos2.xy.y -= collision_height / 2.0f + dy;
+							movement.velocity.x -= (fx/m)*Time::delta_time();
+						}
+					}
+				}
+			}
+
+		}
+
+		pos.xy              += movement.velocity*Time::delta_time();
+		movement.velocity.x += (fx/m)*Time::delta_time();
+		movement.velocity.y += (fy/m)*Time::delta_time();
+		movement.velocity.x = fclampf(movement.velocity.x, -movement.velocity_max.x, movement.velocity_max.x);
+		movement.velocity.y = fclampf(movement.velocity.y, -movement.velocity_max.y, movement.velocity_max.y);
+
+		if (entity == scene.player)
 		{
-        		for(U32 j = 0; j < TILE_COLS; j++)
+			if (scene.dooropen)
 			{
-				Char c = level_manager.level_map[j][i];
-				if(c == 'x')
+				if (pos.xy.x > scene.doorx - 96 && pos.xy.x < scene.doorx + 96 &&
```

Comment:
I will not bug you since the presentation is tomorrow, but for future reference a code with so many hardcoded numbers should raise some red flags. Maybe this needs to be in a file to be loaded with the level or otherwise automatically detectable based on position and size of the entity etc. It's much more important to present something functional tomorrow than it is for it to be general-purpose and pretty, but just for future reference.